_posts/2021-02-01-Dynarec.md

```diff
@@ -0,0 +1,374 @@
+---
+layout: article
+title: Common Dynarec Optimizations
+author: cqcumbers
+---
+
+Many emulator developers are interested in the performance advantages of dynamic
+recompilers, but fewer have implemented one. A good way to start is through one
+of the beginner tutorials available on the Getting Started page. But while the
+simple dynarecs described by these tutorials may be satisfactory for some
+systems, a lot more performance remains on the table, and in between beginner
+content and the cutting edge there are many widely known optimizations that can
+be taken for granted by more experienced developers. This article will cover
+some of these baseline optimizations, with a focus on getting the most
+performance for the least code complexity.
+
+## Introduction
+
+It is important to note before we start that there are some inherent limitations
+to what this guide can help you with. Obviously, it will not cover bottlenecks
+outside of the JIT, of which there are plenty in most full system emulators. As
+you optimize the bottlenecks will naturally shift - remember to profile changes
+and target the parts that actually need improvement, which is frequently
+scheduling and synchronization, peripherals, or graphics rather than actual
+instruction execution - so it probably doesn't make sense to implement all of
+these JIT changes at once, without investigating other components in between.
```

Comment:
You kind of already imply this, but it might be wise to explicit state something along the lines of:

"As with all optimisation tasks, it's important to actually profile your emulator and determine the JIT is currently the bottleneck before proceeding with optimisations"

as a programmer, I know I always forget this and could do with extra explicit reminders from time to time.